rir/src/interpreter/interp.c

```diff
@@ -333,21 +345,56 @@ static SEXP rirCallClosure(SEXP call, SEXP env, SEXP callee, SEXP actuals,
             optimizing = true;
 
             Function* oldFun = fun;
-            SEXP oldBody = body;
-            cp_pool_add(ctx, oldBody);
-            body = globalContext()->optimizer(callee);
+            SEXP oldFunStore = funStore;
+            cp_pool_add(ctx, oldFunStore);
+            funStore = globalContext()->optimizer(funStore);
 
             // TODO: there might be promises with references to the old code!
             // Therefore we keep it around.
             // TODO: first I tried to use R_PreserveObject, but it did not work
             // for large vectors, not sure why, need to investigate
-            cp_pool_add(ctx, body);
+            cp_pool_add(ctx, funStore);
+
+            oldFun->next = funStore;
+            fun = sexp2function(funStore);
+            fun->origin = oldFunStore;
+
+            // Add the new function to the vtable.
+            if (vtable->length < TRUELENGTH(vtableStore)) {
+                vtable->entry[vtable->length++] = funStore;
+            } else {
+                // Allocate a new, larger vtable.
+                DispatchTable* oldVtable = vtable;
+                SEXP oldVtableStore = vtableStore;
+
+                size_t capacity = TRUELENGTH(oldVtableStore) * 2;
+                size_t size = sizeof(DispatchTable) +
+                              (capacity * sizeof(DispatchTableEntry));
+                vtableStore = PROTECT(Rf_allocVector(EXTERNALSXP, size));
+                vtable = sexp2dispatchTable(vtableStore);
+
+                // Initialize the new vtable, copying old entries over.
+                SET_TRUELENGTH(vtableStore, capacity);
+                vtable->magic = DISPATCH_TABLE_MAGIC;
+                vtable->length = oldVtable->length + 1;
+
+                for (size_t i = 0; i < oldVtable->length; i++) {
+                    vtable->entry[i] = oldVtable->entry[i];
+                }
+
+                // Insert the new function version.
+                vtable->entry[oldVtable->length] = funStore;
```

Comment:
uh, if you add the optimized version to the, it won't ever be executed, not?

Reply:
also, shouldn't the length be increased here

Reply:
This uses the old vtable length, the new one is already incremented (l. 379)

Reply:
Yeah, the optimized version is never executed. This is intentional, because I didn't want to wait until we added a new call instruction, and I also didn't want to rip out the optimizer completely.

Maybe I should change the index/increment so that it's a bit clearer that I'm using the old length...

Reply:
> Yeah, the optimized version is never executed

huh, but that's a regression then and has nothing to do with the call instruction, not?

because the old code just replace the existing function with the optimized function and henceforth we'd just use the optimized one.

Reply:
so basically it used to be that the optimized function replaces the existing one.

now the optimized version is just added to the dispatch table, but never used.

Reply:
Correct. The optimized version won't be used until we implement a mechanism for calling a particular version.

I considered reordering the table, so that the optimized version is the first entry. But in general, I think we want the order to be consistent, and have the first entry be the original, most general version.

If you're concerned about the performance regression (do we know how big it would be? We're only optimizing once), I can reorder the table as part of this merged-but-in-progress work, and then change it back (i.e. not change the order when we have a new version) when we have a new call instruction.

Reply:
Well, the baseline version has profiling and tracing enabled. So you should see the same method getting optimized over and over again (since you never replace it by the optimized one)

Reply:
so this is a (1) a perf issue and (2) a testing issue, because we never execute optimized code we will not notice if we break the optimizer

Reply:
Maybe we don't need the unoptimized version? Because the only optimization that is done when compiling a function for the first time is the cleanup pass, and that should always be safe, right?

It is only later in reoptimizeFunction that some more stuff happens (although it is currently just the inliner)

But this could of course change in the future...

Reply:
> So you should see the same method getting optimized over and over again

ah that is probably not true, due to the `!fun->next` test before optimizing. But that's a bit artificial, since prev/next should really go away.

Reply:
> I can reorder the table as part of this merged-but-in-progress work

also I think we might not have the same idea about signatures.

imho the signatures are about things that change the semantics of a version. e.g. the version with the "strict" signature will break when called with other stuff.

but what our "optimizer" currently produces is something which has the same signature as the original version and does not have differing semantics. so for all intents and purposes it is drop in replacement of the source version and I don't see why it should not be used by default (regardless of any future call instructions we have)

Reply:
> It is only later in reoptimizeFunction that some more stuff happens 

I think that is what we are currently talking about, not?

Reply:
Yes, my point was that since we only do cleanup when compiling a new function, we maybe don't need the "unclean" version.

Reply:
Oh my bad, you are right

Reply:
so the way I see it it is kinda orthogonal issue. on the x-axis we have signature and on the y-axis we have optimization level.

When we call a function we pick the most optimized version of the matching signature. Picking the signature must be sound for the whole duration (its not an assumption, because we match it to the actual arguments). So the x-coordinate stays fix.

The different optimization levels on the other hand represent speculative optimizations. If we need to deopt while executing a function we lower the y-coordinate, so to speak.

Note: I am not proposing a 2d dispatch table :) that's just to make it easier to talk about it.

Reply:
We are in agreement. When I say a "mechanism for calling a particular [function] version," I mean that it uses some identifier that distinguishes the signature (or calling convention, but maybe that's too specific) and optimization level, and possibly other metadata. This identifier would be the (x,y) coordinates in your description. (In the call, I might have misspoke and called this identifier the "signature.")

The problem is that this PR is intentionally incomplete. There are these leftovers, such as `origin`, `prev`, and `next`, and the fact that an optimized version is never called.

I could have postponed this PR until all the work is complete (profiling, optimizing for a calling convention, calling a particular version, handling deopt). But then the code is sitting around on the side and getting out of sync (especially since my attention is split across other things). (In fact, there's already a merge conflict.) So I thought it would be better to submit at a reasonable intermediate point.

Now the question is, what is a reasonable intermediate point? I had actually discussed with @janvitek about using the optimized version. I considered swapping the order so that the optimized version is the first (and therefore default) entry, or adding a field with a default offset. But the deoptimizer would also need to be patched, and then all of this reverted when the other things are implemented. Jan suggested it would be simpler to just add the new version to the end of the table and not call it (for now), so that's what I implemented.

Reply:
> But the deoptimizer would also need to be patched

I am certainly fine with submitting an intermediate version. But this is the part I do not understand. If you let prev/origin in place then how come the doptimizer does not work? Sorry, maybe I just don't remember, since I haven't looked at rir myself in a long time. But it seems to me that you should be able to just swap the order and it should work and I fear that if it doesn't that there is a bug.

Reply:
I am not complaining that it's not finished, it's more that I do not understand the scoping. On the one hand you let prev/next in place which is only used by the deoptimizer, but on the other hand you disable optimized code, hence the deoptimizer is not exercised at all.

Reply:
btw, I tried to look at it, but I get

```
failed test rir_regression.R:
   /home/user/Documents/rir/external/custom-r/bin/R -f /tmp/r-test.gzGmu2

R: ../../rir/src/interpreter/interp.c:1382: evalRirCode: Assertion `false && "wrong or unimplemented opcode"' failed.
```

in a debug build.

Reply:
Sorry about the merge, I just wanted to look at the conflict :/
Anyway, the tests pass for me

Reply:
ah crap, I did not update gnur. we really need a solution for this.

Reply:
> But it seems to me that you should be able to just swap the order and it should work and I fear that if it doesn't that there is a bug.

I don't remember if I tried this and it was messy, or if I didn't try it because I wanted to keep it simple.

> I am not complaining that it's not finished, it's more that I do not understand the scoping. On the one hand you let prev/next in place which is only used by the deoptimizer, but on the other hand you disable optimized code, hence the deoptimizer is not exercised at all.

Here's my reasoning: I enabled the optimizer, because I needed a way to grow the vtable. I also didn't want to comment out a bunch of code. I left prev/next because removing them would be extremely invasive. I don't call the optimized code, because it's simpler.

I'll take a look at swapping the order.

@JanJecmen That's OK, I was planning to rebase and mess around with the git history.

Reply:
Ah, I see now. If I rearrange the vtable so that the optimized function is first, then I need to fix deopt.

Before this PR, when a function is optimized, a new version is compiled and added to the linked list. Each version has a pointer to `origin`, the baseline version. The closure is updated to point to the new function version. When a function `f`  is deoptimized, the closure is updated to point to the baseline, `f->origin`.

With this PR, and a local change, new versions are prepended to the vtable, so that the optimized version is the default version that is called. But when a function `f` is deoptimized, yes, `f->origin` is the baseline, but now the closure points to a dispatch table and not a function version. So I can't just update the body of the closure.

I could just swap the order of functions in the vtable again, so that the first entry is now the baseline. But this assumes we only ever have two versions. This is currently true (but may not be in the future), so a more robust solution would involve scanning the vtable to compare pointers with `f->origin` and then making the swap...

I would expect tests to blow up immediately, but this isn't happening, so I'll take a look after lunch...

Reply:
ok, I see

I don't want to block this. If you have a plan in which order to do the steps, then go ahead. I mean the PR is fine, I was just not sure about the scope.

btw. maybe it would make sense to keep slot0 reserved for the "default" method? we don't have to swap the entries, we could just duplicate. eg. initially:  "base, base" and then "opt, base, opt".

Reply:
I'd like to merge as-is (well, after addressing the comment below and rebasing). I don't want to add machinery that I'd remove later.

I don't have a *specific* plan for what to do next, but this is what I have in mind:

1. Generate different versions of a function (maybe optimizations based on some profiling data). This includes saving the "identifier" as some function metadata, which describes the calling convention, optimization level, etc.
2. Add a mechanism for calling a particular function version: given an identifier, scan the dispatch table for that function. Obviously, this will be slow. Also, the deopt mechanism needs to be updated.
3. Improve performance, e.g. with an inline cache

I guess I could swap 1 and 2: only consider "optimization levels" for different versions, and be able to call and deoptimize. Then, after this is done, we can look at other kinds of versions, e.g. calling convention.